src/EditColumnsModal/EditColumnsModal.tsx

```diff
@@ -0,0 +1,104 @@
+import { LfLocalizationService } from "@laserfiche/lf-js-utils";
+import React from "react";
+import { ColumnDef } from '@laserfiche/types-lf-ui-components';
+import './EditColumnsModal.css';
+
+const resources: Map<string, object> = new Map<string, object>([
+    ['en-US', {
+      'NAME': 'Name',
+      'OK': 'Ok',
+      'CANCEL': 'Cancel',
+      'ADD_REMOVE_COLUMNS': 'Add/Remove Columns',
+    }],
+    ['es-MX', {
+      'NAME': 'Name -Spanish',
+      'OK': 'Ok - Spanish',
+      'CANCEL': 'Cancel - Spanish',
+      'ADD_REMOVE_COLUMNS': 'Add/Remove Columns - Spanish',
+    }]
+]);
+
+export default class EditColumnsModal extends React.Component<{onClose: (columns?: ColumnDef[]) => Promise<void>; errorMessage: string; initialColumnsSelected: ColumnDef[];allColumnOptions: ColumnDef[] }, {columnsSelected: ColumnDef[]; open: boolean}> {
+
+    localizationService: LfLocalizationService = new LfLocalizationService(resources);
+    closeOnEscapeKeyDown = (e:KeyboardEvent) => {
+        if ((e.code === 'Escape')) {
+            this.props.onClose();
+        }
+    };
+
+    constructor(props: any) {
+        super(props);
+        this.state = {
+            columnsSelected: this.props.initialColumnsSelected,
+            open: false
+
+        };
+        console.log(this.state.columnsSelected);
+    }
+
+    componentDidMount() {
+        document.body.addEventListener("keydown", this.closeOnEscapeKeyDown);
+        const container = document.getElementById('edit-columns-dialog-content');
+        container?.addEventListener("focusout", (ev)=>{
+          if (!container.contains(ev.relatedTarget as Node)) 
+          document.getElementById('edit-columns-dialog-modal-close-dialog')?.focus();
+        });
+        setTimeout(() => {
+            this.setState({open: true});
+          document.getElementById('edit-columns-name-input')?.focus();
+        });
+      }
+  
+  
+      closeDialog = (columns?: ColumnDef[]) => {
+          this.props.onClose(columns);
+      };
+
+      onCheckboxChange(column: ColumnDef) {
+        // we provide a new array (not just new values) to trigger change detection
+        let newSelected: ColumnDef[];
+        if (this.state.columnsSelected.includes(column)) {
+            newSelected = this.state.columnsSelected.filter(c => c != column).map(value => value);

```

Comment:
what does  .map(value => value) do?

Reply:
this can be simplified. We can pass the onChange event to the onCheckboxChange function, and from the event detail we can tell if this current check box is checked or not. If it is checked, we add it to columnsSelected. If not, we remove it from columnsSelected

Reply:
it came up briefly during mobbing, but I was using .map(value => value) to create a shallow copy so that react would detect changes when I reassigned the array. Not sure if there's a more transparent way to do this. I saw Alex mentioned {...object} as an alternative? 

I think the rest of your suggestions is basically what I'm already doing, except that I'm not passing the event to the handler.